### ISSUE TITLE
Best Practice to implement Postgres Row Level Security

### ISSUE THREAD

**Issue by pratikpparikh** [koskimas/kysely#330]
What is the best way to implement Postgres Row Level Security using Kysely?  Do you have a concept of interceptor in Kysely?  Is WithSchemaPlugin the mechanism to do this?  If so please provide a example.

Thank you in advance.

Regards,
Pratik Parikh

**Comment by koskimas** [comment]
I have never done anything using postgres row level security, so can't really help you. There's nothing built-in for that in Kysely. We don't have interceptors. You could maybe do achieve something useful with plugins, but if I understood correctly, you want to run extra queries that set the role. That won't be possible using plugins, since the transform hook is not async.

You can see a bunch of examples in the codebase. For example the `WithSchemaPlugin` you just mentioned.

**Comment by igalklebanov** [comment]
@lancejpollard check https://github.com/kysely-org/awesome-kysely?tab=readme-ov-file#plugins, there are some security-related plugins there, maybe they'll help.

Kysely is an unopinionated query builder. Nothing more, nothing less. Security patterns are out of scope for Kysely's core and belong in community libraries.

Follow the official guidelines https://www.postgresql.org/docs/current/ddl-rowsecurity.html.

**Comment by Gaspero** [comment]
> What should I do here exactly?
> 
> Maybe having a Kysely recommendation for this situation might be useful (to myself and others).

@lancejpollard  You can find an example of RLS policies at Supabase docs https://supabase.com/docs/guides/auth/row-level-security#policies

Like Igal mentioned you can`t use Kysely to build such query. But you could execute such query with Kysely. 
See an example https://kyse.link/?p=s&i=0kjyO4eBSXFs8rhK7sbW

I guess you can take this example even further if you create a helper https://kysely.dev/docs/recipes/extending-kysely

**Comment by stoicskyline** [comment]
> PS: AFAIK the warning is only applied to requests made using PostgREST API. If you are are not using PostgREST API and haven't leaked your Anon key in any client app - you have nothing to worry about. External user might be able to call PostgREST API using Anon key. If no key is passed to request, it will be treated as unauthorized.

To clarify: `anon` key is fine in the client since it's used by `@supabase/supabase-js` to do anonymous PostgREST requests, but do **not** leak the service role key in the client.

**Comment by stoicskyline** [comment]
If you already chose Supabase, then the path of least resistance may be to use their tooling where possible, and opt in Kysely/etc for other aspects.  

Supabase has a huge feature set that they aspire to create, but it may not all be at the same quality e.g. the raw SQL may not be easy for junior devs.  However, the more additional libs you add, the higher your bus count.

You'll need consider if each lib will live longer, be more maintainable for your team size, etc for different aspects:
* `supabase-js` or another solution involving Kysely/etc 
* RLS
* migrations
* seeding
* raw SQL queries - which is not easy to do https://github.com/orgs/supabase/discussions/3419

The last one it seems you have to roll Kysely or another lib to do.

**Comment by giorgiogross** [comment]
Kysely actually supports RLS out of the box, this is what I'm doing:
```
		// Cache the pool somewhere to be reused across connections
		const pool = new pg.Pool({
			// connectionString: '...you connection string...'
		});

		/*
                Then initialise dialect.pool, overriding the connect and end params as below.
		Ideally, you'd do this in some kind of factory that instantiates a new Kysely instance
		for each request, but reuses the cached pool.
                "vars" here is the request context variables passed by my gateway (eg express, hono, ...)
                which were initialised when authenticating incoming jwt bearer tokens.
                */
		const dialect = new PostgresDialect({
			pool: {
				connect: async () => {
					const client = await pool.connect();

					/*
          				Set JWT claims to enable RLS, mirroring what Supabase does.
          				See https://github.com/drizzle-team/drizzle-orm/issues/594#issuecomment-2016607868
           				*/
					await client.query(
						"SELECT set_config('request.jwt.claims', $1, FALSE)",
						[JSON.stringify(vars.jwtPayload)]
					);
					await client.query(
						"SELECT set_config('request.jwt.claim.sub', $1, FALSE)",
						[vars.userId]
					);
					await client.query(
						"SELECT set_config('request.jwt.claim.email', $1, FALSE)",
						[vars.userEmail]
					);

					// Only support authenticated for authed users:
					await client.query(
						"SELECT set_config('request.jwt.claim.role', 'authenticated', FALSE)"
					);
					await client.query('SET ROLE authenticated');
					return client;
				},
				end: async () => pool.end(),
			},
		});

                // Then use the db with the initialised dialect. Pass in your own types.
		const db = new Kysely<Database>({
			dialect,
		});
```

Was taking in https://github.com/brianc/node-postgres/issues/2897 and https://github.com/drizzle-team/drizzle-orm/issues/594#issuecomment-2016607868

Every time a new client is used, `connect()` is invoked which initialises the client's config with `set_config(...)` and `SET ROLE`. You simply have to make sure that you don't reuse a client which was set up for another user, which I'm getting around with by instantiating Kysely for each request and just caching the pool.

**Comment by bombillazo** [comment]
Hey @giorgiogross , how are you creating a new client on every request? Do you have a function that returns a `new Kysely()` instance?

**Comment by bombillazo** [comment]
Hey @giorgiogross Another detail: For example, Supabase exposes 2 ports for different [connection methods](https://supabase.com/docs/guides/database/connecting-to-postgres#choosing-a-connection-method): one for `session` connections and another for `transaction` connections. Not sure which type of connection you tried for your setup, but I tried this approach and (at least using the transaction method) the `connect` logic will run _for each_ SQL query sent to the database.

The dialect object has a function called `onCreateConnection` and using this, I see the logic runs only once per request.

**Comment by ardabeyazoglu** [comment]
[this article](https://arda.beyazoglu.com/supabase-backend-custom-query-security-with-rls)
 might help, especially for anyone using kysely with supabase.

**Comment by bombillazo** [comment]
@ardabeyazoglu thanks for sharing.

It seems this is a common pitfall of using backend DB connections within the same request, but rarely does all the logic to an API endpoint reside within the same function transaction context. We want a way to have a client per request and use that client for all logic where a DB connection is required. We are using the transaction mode SB link. Im not so sure how Kysely works but Im guessing each query/call to the database uses an individual connection, so in transaction mode it will fetch a connection from the pool for each query executed? or does the Kysely client use the same pool connection for all the calls made with that client?

I ask because if it is the latter, we can use helper functions that change the role where needed without needing a transaction, which greatly simplifies being able to break our db code into different services and call functions from different places/handlers/services. If it is the former, we're almost obligated to wrap everything in a transaction to ensure any changes to roles is withing the context of a single transaction query that uses the same connection and does not interfere with other connections or the pool will not mix up connections and their context/role changes across different requests.

Im still experimenting and trying things out, I hope to find a solution that is flexible enough to no having to inject around a client transaction but still secure and avoid leaking roles.

**Comment by bombillazo** [comment]
Just to give an example, we have a cache service that should only be accessible from our backend, and we use a service/postgres role to access this. we use Kysely to run `get` and `set` functions that manipulate the cache.

We want to use these functions ad hoc, and internally they ideally take care of using a client connection with the proper role and making the change. But perhaps to be 100% safe we always need to pass the client to ensure no other client connection is made.

**Comment by ardabeyazoglu** [comment]
@bombillazo i think you misunderstood my article. I dont recommend putting all queries into single transaction. Use it whenever you need, putting each query into separate transactions if you want, does not matter. As long as, you set transaction level role and claims, it will be ok. In your case, you should pass kysely client to your caching system and it must use transactions like i recommended in my article. Using transactions is the only safe way to use connection pools and also the most performant method. Splitting up your queries to different transactions, or abstracting away this process depends on your application and frameworks. We use nestjs, and creating a global, singleton database service that sends the queries with transactions and roles (as in the araticle) and it works perfect.

**Comment by bombillazo** [comment]
@ardabeyazoglu  I meant putting all queries related to a single endpoint logic (for example, if we need to select, update and insert as part of a logic for a single endpoint) then we need to have all those queries part of the same transaction to ensure they run with the same connection and there are no role change issues.

Like I mentioned we may have utils and services that do DB selects or updates as well, so we're ending up having to inject the transaction client in order to ensure those logic steps run with he same connection.

**Comment by bfelbo** [comment]
@giorgiogross Thanks for sharing that code example, very helpful. Note that you're currently making multiple DB roundtrips to authenticate every request. You can set multiple configs in a single request, the PostGraphile library explains it nicely here :)
https://www.graphile.org/postgraphile/security/#how-it-works

**Comment by bfelbo** [comment]
> You simply have to make sure that you don't reuse a client which was set up for another user, which I'm getting around with by instantiating Kysely for each request

Do you have an idea of how much overhead this adds?

I'd imagine it wouldn't be too hard to add support in Kysely to avoid this if it's expensive.

**Comment by Dudeonyx** [comment]
> [this article](https://arda.beyazoglu.com/supabase-backend-custom-query-security-with-rls) might help, especially for anyone using kysely with supabase.

[updated link](https://blog.ardabeyazoglu.com/supabase-backend-custom-query-security-with-rls) , had a mini heart attack when my bookmark no longer worked lol, ty
